### DISCUSSION TITLE
Ant 连接服务器

### DISCUSSION THREAD

**Discussion by julywind168** [ejoy/ant#72] [Q&A] [answered]
我看 wiki 中有 [网络连接服务](https://github.com/ejoy/ant/wiki/ActorModel#%E7%BD%91%E7%BB%9C%E8%BF%9E%E6%8E%A5%E6%9C%8D%E5%8A%A1)
但是我并没有找到，ant.net 下面的服务是没放出来吗？我用现有的 ant.net 尝试连接服务器 （代码里没有暴露 connect api, 我手动加上了）发现出现了EINPROGRESS，原因是 bee.socket 默认都是非阻塞的, 我又改了 c 代码，在 socket("tcp", block) 可以设置是否阻塞，改了之后连接是成功了，但是游戏界面卡住了。
也许连接成功后再将 fd 设置成非阻塞就可以用了，但是我没有尝试。因为我看引擎内用了 connect api，难道ios上没问题，mac上就有问题吗。

**Comment by actboy168** [comment]
对于bee.socket, tools/fileserver有相关的使用和封装。

然后永远不要有阻塞型socket。

**Reply by julywind168** [reply]
🤔那连接服务器的正确姿势是啥样的呢

**Reply by cloudwu** [reply]
现在网络连接服务是合并在 io 服务中的，因为一个进程不需要两个处理网络的地方。ant.net 会通过动态 patch 的方式扩展 io 服务。

现在 ant.net 在 ant.webserver  中有用到，accept 是可以用的。connect 没暴露出来，因为暂时没测试。理论上应该是差不多的。你可以调试一下，如果有 bug 可以提个 pr 。

但不应该改成阻塞模式，目前 webserver 也工作正常，应该不关阻塞模式的事。

**Reply by julywind168** [reply]
好的，我再看看～

**Reply by cloudwu** [reply]
非阻塞模式下， connect 后应该监听可写事件，收到可写事件后就和 accept 的 socket 一样工作了。

我看了下代码，是按这个逻辑写的。https://github.com/ejoy/ant/blob/master/pkg/ant.net/network.lua#L105-L113
但是没有实测过，我们的游戏没有网络部分。你可以看看到底为什么有  EINPROGRESS 错误。主要看，SELECT_WRITE 事件是否能阻塞住代码执行流程，正确的行为是阻塞在 `ltask.wait(obj.wt)` 这行。

是否在 SELECT_WRITE 事件前就去读写 fd 了。

**Comment by julywind168** [comment, marked as answer]
我修复了一下 network，见 #73 ，后面的读写还没测试

**Reply by cloudwu** [reply]
不错。

这个 ant.net 现在比较粗糙。毕竟客户端的网络性能不太重要。不过我设想，如果客户端固定了封包协议，真实应用中可以考虑把协议解析放在 io 服务中，而不是通过类 socket api 这样直接解析数据流。

**Reply by julywind168** [reply]
可以的，封装一下肯定让 api 更好用，不然每个项目都得自己去实现一遍
